airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StagingSqlOperations.java

```diff
@@ -0,0 +1,202 @@
+    return nameTransformer.applyDefaultCase(String.format("%s/%s/%s/%02d/%02d/%02d/",
+        connectionId,
+        getStageName(namespace, streamName),
+        writeDatetime.year().get(),
+        writeDatetime.monthOfYear().get(),
+        writeDatetime.dayOfMonth().get(),
+        writeDatetime.hourOfDay().get()));
+  }
+
+  @Override
+  public void insertRecordsInternal(final JdbcDatabase database,
+                                    final List<AirbyteRecordMessage> records,
+                                    final String schemaName,
+                                    final String path) {
+    final List<Exception> exceptionsThrown = new ArrayList<>();
+    boolean succeeded = false;
+    while (exceptionsThrown.size() < UPLOAD_RETRY_LIMIT && !succeeded) {
+      try {
+        loadDataIntoStage(database, path, records);
+        succeeded = true;
+      } catch (final Exception e) {
+        LOGGER.error("Failed to upload records into stage {}", path, e);
+        exceptionsThrown.add(e);
+      }
+      if (!succeeded) {
+        LOGGER.info("Retrying to upload records into stage {} ({}/{}})", path, exceptionsThrown.size(), UPLOAD_RETRY_LIMIT);
+        // Force a reconnection before retrying in case error was due to network issues...
+        s3Client = s3Config.resetS3Client();
+      }
+    }
+    if (!succeeded) {
+      throw new RuntimeException(String.format("Exceptions thrown while uploading records into stage: %s", Strings.join(exceptionsThrown, "\n")));
+    }
+  }
+
+  private void loadDataIntoStage(final JdbcDatabase database, final String stage, final List<AirbyteRecordMessage> records) {
+    final long partSize = s3Config.getFormatConfig() != null ? s3Config.getFormatConfig().getPartSize() : DEFAULT_PART_SIZE;
+    final String bucket = s3Config.getBucketName();
+    final String objectKey = String.format("%s%s", stage, UUID.randomUUID());
+    final StreamTransferManager uploadManager = S3StreamTransferManagerHelper
+        .getDefault(bucket, objectKey, s3Client, partSize)
+        .checkIntegrity(true)
+        .numUploadThreads(DEFAULT_UPLOAD_THREADS)
+        .queueCapacity(DEFAULT_QUEUE_CAPACITY);
+    boolean hasFailed = false;
+    try {
+      final List<MultiPartOutputStream> outputStreams = uploadManager.getMultiPartOutputStreams();
+      try (final MultiPartOutputStream outputStream = outputStreams.get(0)) {
+        try (final CSVPrinter csvPrinter = new CSVPrinter(new PrintWriter(outputStream, true, StandardCharsets.UTF_8), CSVFormat.DEFAULT)) {
+          final CsvSheetGenerator csvSheetGenerator = new StagingDatabaseCsvSheetGenerator();
+          createStageIfNotExists(database, stage);
+          for (final AirbyteRecordMessage recordMessage : records) {
+            final var id = UUID.randomUUID();
+            csvPrinter.printRecord(csvSheetGenerator.getDataRow(id, recordMessage));
+          }
+        }
+      }
```

Comment:
Nitpick. You can define multiple auto-closable objects with in the `try` statements to avoid having nested `try` blocks. I think one `try` statement will be enough:

```suggestion
    try (final MultiPartOutputStream outputStream = uploadManager.getMultiPartOutputStreams().get(0);
        final CSVPrinter csvPrinter = new CSVPrinter(new PrintWriter(outputStream, true, StandardCharsets.UTF_8), CSVFormat.DEFAULT)) {
      final CsvSheetGenerator csvSheetGenerator = new StagingDatabaseCsvSheetGenerator();
      createStageIfNotExists(database, stage);
      for (final AirbyteRecordMessage recordMessage : records) {
        final var id = UUID.randomUUID();
        csvPrinter.printRecord(csvSheetGenerator.getDataRow(id, recordMessage));
      }
    } catch (final Exception e) {
```